go-pray: init at 0.1.13

0xzer0x opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

0xzer0x committed:
maintainers: add _0xzer0x

0xzer0x committed:
go-pray: init at 0.1.13

r-vdp commented on pkgs/by-name/go/go-pray/package.nix:
```diff
@@ -0,0 +1,39 @@
+  ldflags = [
+    "-X 'github.com/0xzer0x/go-pray/internal/version.version=${finalAttrs.version}'"
+    "-X 'github.com/0xzer0x/go-pray/internal/version.buildCommit=${finalAttrs.version}'"
+    "-X 'github.com/0xzer0x/go-pray/internal/version.buildTime=1970-01-01T00:00:00Z'"
```
Using SOURCE_EPOCH would probably be more useful here. See https://reproducible-builds.org/docs/source-date-epoch/

But we can also just not set it at all, I guess?

r-vdp commented on pkgs/by-name/go/go-pray/package.nix:
```diff
@@ -0,0 +1,39 @@
+  ];
+  ldflags = [
+    "-X 'github.com/0xzer0x/go-pray/internal/version.version=${finalAttrs.version}'"
+    "-X 'github.com/0xzer0x/go-pray/internal/version.buildCommit=${finalAttrs.version}'"
```
Is this useful? We're setting the commit to a version string. You can actually get the commit hash, but I wouldn't recommend it, it's a bit messy.

r-vdp commented on pkgs/by-name/go/go-pray/package.nix:
```diff
@@ -0,0 +1,65 @@
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-wwTAwiO1XONxd6NPXKjzu5B7R211xAgBGPPAvQwOp8I=";
+    # NOTE: Required for setting version commit and build time
+    leaveDotGit = true;
```
What's the added value of having the commit hash? Because we try to avoid doing this unless really required.

r-vdp replied:
Can't we just provide the version without the commit hash?

0xzer0x replied:
Alright, I removed it as requested.

0xzer0x commented:
I can use the `SOURCE_DATE_EPOCH` for now as a placeholder in order not to break the `version` sub-command. However as pointed out by #256270 it is not set correctly, currently it evaluates to `Tue Jan 01 00:00:00 1980` when used.

0xzer0x committed:
go-pray: refactor to remove build commit and time workaround

This approach of keeping the required metadata in the actual package.nix
maintains reproducibility while keeping metadata required by upstream
package.
